### PR TITLE
Initial Production version of the SEFT consumer

### PR DESCRIPTION
**Pull request by warrenbailey**
__Description__

Initial version of the Seft Consumer service. This service reads encrypted and signed JWT tokens from the Seft.Responses queue. Inside the JWT is a base64 encoded version of the SEFT excel file. The service reconstructs the file and transmits it over FTP to an endpoint inside the ONS.

__How to Test__
There are several unit tests which invoke the functionality. However there is also an end to end integration test which uses a real rabbit mq server to prove the process, this is disabled by default.

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from sdx.common.logger_config import logger_initial_config
+
+from app import settings
+
+__version__ = "1.0.0"
```
Comment (necrophonic) [inline app/__init__.py]: Being pedantic it should probably be `0.1.0` until we actually want to release

File: app/consumer.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+from app import settings
+from app.decrypter import Decrypter
+from app.sdxftp import SDXFTP
+
+logger = wrap_logger(logging.getLogger(__name__))
+
+'''
+The AsyncConsumer will automatically pick up the RABBIT queue from a file called settings with var RABBIT_QUEUE
+TODO this really needs to change
```
Comment (necrophonic) [inline app/consumer.py]: Yup it does 😂

File: app/consumer.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        super().__init__(log)
+        self._decrypter = Decrypter(settings.RAS_SEFT_PUBLIC_KEY,
+                                    settings.SDX_SEFT_PRIVATE_KEY,
+                                    settings.SDX_SEFT_PRIVATE_KEY_PASSWORD)
+
+        self._ftp = SDXFTP(logger, settings.FTP_HOST, settings.FTP_USER, settings.FTP_PASS, settings.FTP_PORT)
+
+    def on_message(self, unused_channel, basic_deliver, properties, body):
+        encrypted_jwt = body.decode("utf-8")
+        logger.debug("Message Received")
```
Comment (necrophonic) [inline app/consumer.py]: Should be `info` - we're always going to want to log the receipt of a message (ideally with an identifier [like `tx_id` if such a thing exists])

File: app/consumer.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import logging
+
+from sdx.common.async_consumer import AsyncConsumer
+from structlog import wrap_logger
+
+from app import settings
+from app.decrypter import Decrypter
+from app.sdxftp import SDXFTP
+
+logger = wrap_logger(logging.getLogger(__name__))
```
Comment (necrophonic) [inline app/consumer.py]: Should this be using json format now?

File: app/consumer.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        decrypted_payload = self._decrypter.decrypt(encrypted_jwt)
+
+        file_contents = decrypted_payload.get("file")
+        file_name = decrypted_payload.get("filename")
+        logger.debug("Decrypted file", file_name=file_name)
+
+        decoded_contents = base64.b64decode(file_contents)
+
+        with open("./seft_files/" + file_name, "wb") as fb:
```
Comment (necrophonic) [inline app/consumer.py]: Is this just for testing? We shouldn't be writing to the file system at all.

File: app/consumer.py (hunk 3 above)
Comment (necrophonic) [inline app/consumer.py]: Should also then bind the identifier to the logger for tracability

File: app/decrypter.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+IV_EXPECTED_LENGTH = 12
+CEK_EXPECT_LENGTH = 32
+
+
+# TODO this is a copy of the code in sdx-decrypt, we need to change this asap and either shift the code into sdx-common or
+# allow decrypt to process both EQ and RAS files (preferable)
```
Comment (necrophonic) [inline app/decrypter.py]: 👍

File: app/consumer.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                                    settings.SDX_SEFT_PRIVATE_KEY,
+                                    settings.SDX_SEFT_PRIVATE_KEY_PASSWORD)
+
+        self._ftp = SDXFTP(logger, settings.FTP_HOST, settings.FTP_USER, settings.FTP_PASS, settings.FTP_PORT)
+
+    def on_message(self, unused_channel, basic_deliver, properties, body):
+        encrypted_jwt = body.decode("utf-8")
+        logger.debug("Message Received")
+
+        decrypted_payload = self._decrypter.decrypt(encrypted_jwt)
```
Comment (necrophonic) [inline app/consumer.py]: What happens when decryption fails?

File: app/consumer.py (hunk 5 above)
Comment (warrenbailey) [inline app/consumer.py]: That was just me testing, i'll remove

File: app/__init__.py (hunk 1 above)
Comment (warrenbailey) [inline app/__init__.py]: Yeah probably should, i'll change, though i think we need a discussion on our semantic versioning. Something for the next tech meeting?

File: app/consumer.py (hunk 4 above)
Comment (warrenbailey) [inline app/consumer.py]: Yes, i'll change

File: app/consumer.py (hunk 3 above)
Comment (warrenbailey) [inline app/consumer.py]: Agreed, will change

File: app/consumer.py (hunk 7 above)
Comment (warrenbailey) [inline app/consumer.py]: Need a quarantine queue or something

File: app/consumer.py (hunk 4 above)
Comment (warrenbailey) [inline app/consumer.py]: Though maybe that is something which should be in the common library?

File: app/__init__.py (hunk 1 above)
Comment (warrenbailey) [inline app/__init__.py]: Done

File: app/consumer.py (hunk 4 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: app/consumer.py (hunk 3 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: app/consumer.py (hunk 7 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: app/consumer.py (hunk 5 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,2 +1,3 @@
 ### Unreleased
   - Add codacy badge
```
Comment (necrophonic) [inline CHANGELOG.md]: Not yours, but may be worth removing this line - it tends to imply an amendment on something that hasn't been released

File: README.md [hunk 9]
```diff
@@ -2,10 +2,19 @@
...
 
-Microservice for consuming SEFT files from RAS and sending to ONS internal
-
 ### License
 
 Copyright ©‎ 2016, Office for National Statistics (https://www.ons.gov.uk)
 
-Released under MIT license, see [LICENSE](LICENSE) for details.
\ No newline at end of file
+Released under MIT license, see [LICENSE](LICENSE) for details.
```
Comment (necrophonic) [inline README.md]: License should be down the bottom of the readme - the overview and getting started sections should be at the top. (actually shouldn't need "overview" as a heading - should just be the first line after the title)

File: README.md [hunk 10]
```diff
@@ -2,10 +2,19 @@
...
+
+### Overview
+Microservice for consuming SEFT files from RAS via a Rabbit MQ and sending onto the internal ONS network
+
+# Generate the public/private key pair
```
Comment (necrophonic) [inline README.md]: Could do with a proper "Getting started" section and a block covering env vars (like https://github.com/ONSdigital/sdx-receipt-rrm)

File: app/consumer.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+        delivery_count = self.get_delivery_count_from_properties(properties)
+
+        tx_id = self.get_tx_id_from_properties(properties)
+        if not tx_id:
+            self.quarantine_publisher.publish_message(body)
+            self.reject_message(basic_deliver.delivery_tag)
+            logger.error("No tx_id so quarantining message",
+                         action="quarantined",
+                         tx_id=tx_id,
```
Comment (necrophonic) [inline app/consumer.py]: If there's no `tx_id` then you can't log the `tx_id`

File: app/consumer.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    @staticmethod
+    def get_tx_id_from_properties(properties):
+        """
+        Returns the tx_id for a message from a rabbit queue. The value is
+        auto-set by rabbitmq.
+        """
+        tx_id = None
+        try:
+            if properties.headers:
+                tx_id = properties.headers['tx_id']
```
Comment (necrophonic) [inline app/consumer.py]: What if there are headers but those headers don't contain a `tx_id`?

File: app/consumer.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+    def __init__(self, log=None):
+        super().__init__(log)
+        self._decrypter = Decrypter(settings.RAS_SEFT_PUBLIC_KEY,
+                                    settings.SDX_SEFT_PRIVATE_KEY,
+                                    settings.SDX_SEFT_PRIVATE_KEY_PASSWORD)
+
+        self.quarantine_publisher = QueuePublisher(logger, settings.RABBIT_URLS, settings.RABBIT_QUARANTINE_QUEUE)
+
+        self._ftp = SDXFTP(logger, settings.FTP_HOST, settings.FTP_USER, settings.FTP_PASS, settings.FTP_PORT)
```
Comment (JamesGardiner) [inline app/consumer.py]: Line breaks here

File: app/settings.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+def get_key(key_name):
+    key = open(key_name, 'r')
+    contents = key.read()
+    return contents
+
+LOGGING_LEVEL = os.getenv("LOGGING_LEVEL", "DEBUG")
+
+RABBIT_URL = 'amqp://{user}:{password}@{hostname}:{port}/{vhost}'.format(
+    hostname=os.getenv('RABBITMQ_HOST', 'localhost'),
```
Comment (necrophonic) [inline app/settings.py]: Should probably try to get away without defaults as this is a fresh service

File: app/consumer.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                         action="nack",
+                         exception=e,
+                         tx_id=tx_id,
+                         delivery_count=delivery_count)
+
+        file_contents = decrypted_payload.get("file")
+        file_name = decrypted_payload.get("filename")
+        logger.debug("Decrypted file", file_name=file_name, tx_id=tx_id)
+
+        decoded_contents = base64.b64decode(file_contents)
```
Comment (JamesGardiner) [inline app/consumer.py]: Could `file_contents` and `file_name` potentially be `None`? If so, this needs a try/except block.

File: app/consumer.py [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+        try:
+            self._ftp.deliver_binary(settings.FTP_FOLDER, file_name, decoded_contents)
+            logger.debug("Delivered to FTP server", tx_id=tx_id)
+            self.acknowledge_message(basic_deliver.delivery_tag)
+        except IOError as e:
+            logger.exception("Unable to deliver to the FTP server", e)
+            self.nack_message(basic_deliver.delivery_tag, tx_id=tx_id)
+
+    @staticmethod
```
Comment (JamesGardiner) [inline app/consumer.py]: `@staticmethods` should come before `__init__` method.

File: app/consumer.py (hunk 13 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: CHANGELOG.md (hunk 8 above)
Comment (warrenbailey) [inline CHANGELOG.md]: Done

File: README.md (hunk 9 above)
Comment (warrenbailey) [inline README.md]: Done

File: app/consumer.py (hunk 11 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: app/consumer.py (hunk 16 above)
Comment (warrenbailey) [inline app/consumer.py]: Done

File: app/consumer.py (hunk 15 above)
Comment (warrenbailey) [inline app/consumer.py]: Yes, this code needs hardening. I'll sort it

File: README.md (hunk 10 above)
Comment (warrenbailey) [inline README.md]: I'll add

File: app/settings.py (hunk 14 above)
Comment (warrenbailey) [inline app/settings.py]: These are needed by the unit tests, so removing them breaks the tests.

We need to convert this to a flask app in order to have a health check. Once we do that we can use flask config and remove this settings file completely, hence removing the default values

File: app/consumer.py (hunk 12 above)
Comment (warrenbailey) [inline app/consumer.py]: It'll raise a KeyError and get caught in the except, logging that it's missing a tx_id

File: app/settings.py (hunk 14 above)
Comment (JamesGardiner) [inline app/settings.py]: Whilst keeping a configuration for testing?

File: app/consumer.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+            logger.error("Failed to process",
+                         action="nack",
+                         exception=e,
+                         tx_id=tx_id,
+                         delivery_count=delivery_count)
+
+        file_contents = decrypted_payload.get("file")
+        file_name = decrypted_payload.get("filename")
+        if not file_contents:
+            self.quarantine_publisher.publish_message(body)
```
Comment (JamesGardiner) [inline app/consumer.py]: I think this would be better as a try/except block

Comment (warrenbailey) [inline app/consumer.py]: OK, i'll change

File: app/settings.py (hunk 14 above)
Comment (warrenbailey) [inline app/settings.py]: We have to have test config some where

**Pull request merged**